### PR TITLE
Add vertex normals

### PR DESCRIPTION
**Pull request by abhimadan**
With vertex normals, we get Phong shading for free, so we can now represent smooth surfaces in `WorkingGeometry`.

There were a few things I found that I think were incorrect before this PR:
- `WorkingGeometry` assumed that normals corresponded to faces in the WebGL buffers, but they actually correspond to vertices.
- `WorkingGeometry.scale` has been extended to work with non-collinear pull and destination points. The scale factors were also not being computed properly before.
- WorkingGeometry.scaleByFactor has been fixed to produce invertible matrices. There were cases before where the scale could be 0 (e.g., if pullPoint == destinationPoint).

Some of these changes have resulted in some functions being removed from WorkingGeometry.

### CODE REVIEW

File: src/geometry/WorkingGeometry.ts [hunk 1]
```diff
@@ -143,22 +148,60 @@ export class WorkingGeometry {
...
+                vec3.normalize(vec3.create(), translatedDestinationPoint)
+            )
+        );
+
+        const moveOriginToHold = mat4.translate(mat4.create(), mat4.create(), holdPoint);
+
+        // Assemble all the matrices!
+        // Note that matrix is modified after each mul call, but the linter can't figure that out,
+        // so the const is only there to appease the linter
+        const matrix = moveOriginToHold;
```
Comment (davepagurek) [inline src/geometry/WorkingGeometry.ts]: `const` only means a const reference, and since we're modifying the contents of the variable without changing the reference, `const` still makes sense.

**Review by davepagurek — APPROVED** [review]
Thanks for doing those fixes!

**Review by armcburney — APPROVED** [review]
LGTM! Nice work! 👍

File: src/geometry/WorkingGeometry.ts [hunk 2]
```diff
@@ -143,22 +148,60 @@ export class WorkingGeometry {
...
+        mat4.mul(matrix, matrix, rotXAxisToDestAxis);
+        mat4.mul(matrix, matrix, scalePullToDestOnXAxis);
+        mat4.mul(matrix, matrix, rotPullToXAxis);
+        mat4.mul(matrix, matrix, moveHoldToOrigin);
```
Comment (tammy) [inline src/geometry/WorkingGeometry.ts]: I'm having a hard time following what each of these matrices do, and why they're necessary visually.

I was under the impression that the function `fromRotationTranslationScaleOrigin` allowed us to no longer have to specify each matrix and multiply them together (e.g., no need to translate to origin and then rotate). The error you mentioned before I thought could be fixed in `getScalingVector` because it was scaling to 0 instead of 1 in certain cases.

I'd rather we still use `fromRotationTranslationScaleOrigin` in order to have parity with `scaleByFactor` since they're essentially supposed to be doing the same things, and that function actually fits the purpose of `scale` much better than `scaleByFactor` so I think it's possible. This is assuming that it can still be done without witchcraft and black magic 😃

File: src/geometry/WorkingGeometry.ts [hunk 3]
```diff
@@ -168,20 +211,13 @@ export class WorkingGeometry {
...
      * @param {vec3} holdPoint: point to scale from, default to true origin.
      */
-    public scaleByFactor(factor: number, pullPoint: vec3, holdPoint: vec3 = vec3.create()) {
-        const scalingDirection: vec3 = vec3.sub(vec3.create(), pullPoint, holdPoint);
-        const absScalingDirection: vec3 = vec3.fromValues(
-            Math.abs(scalingDirection[0]),
-            Math.abs(scalingDirection[1]),
-            Math.abs(scalingDirection[2])
-        );
+    public scaleByFactor(factor: number, holdPoint: vec3 = vec3.create()) {
```
Comment (tammy) [inline src/geometry/WorkingGeometry.ts]: Why is there no longer a `pullPoint`? The shape can be scaled by a factor but it has to scale to a certain direction, unless we want to define it as always scaling to the positive quadrant and then move it later if needed.

File: tests/geometry/WorkingGeometry.spec.ts [hunk 4]
```diff
@@ -324,19 +340,24 @@ describe('WorkingGeometry', () => {
 
             expect(square.vertices).toEqualArrVec4(TestHelper.square().vertices);
         });
-        it('will throw an error when it is skewed', () => {
+        it('will perform a rotation when the pull and destination points are not colinear', () => {
```
Comment (tammy) [inline tests/geometry/WorkingGeometry.spec.ts]: I'm seeing a rotation and translation as well? The square origin (0, 0, 0) translated to (1, 1, 0) and then rotated -90 degrees on the z-axis about (1, 1, 0). I feel like this isn't expected behaviour, I was expecting the scale to flatten the square to a line with 4 points with 2 each on (1, 0, 0) and (0, 0, 0).

Can you help me visualize this?

**Review by tammy — CHANGES_REQUESTED** [review]

File: src/geometry/WorkingGeometry.ts (hunk 1 above)
Comment (abhimadan) [reply src/geometry/WorkingGeometry.ts]: Got it. I'll remove that comment.

File: src/geometry/WorkingGeometry.ts (hunk 2 above)
Comment (abhimadan) [reply src/geometry/WorkingGeometry.ts]: So the order of `mat4.mul` multiplies the first operand by the second one, so this block actually needs to be read from the bottom up, unfortunately. This is the way I'm visualizing what this transformation is doing:
1) Apply a translation so that `holdPoint` is the origin.
2) Apply a rotation so that `pullPoint` is on the x-axis.
3) Apply a scale along the x-axis so that `pullPoint` has the same distance from `holdPoint` (which is currently the origin) as `destinationPoint`.
4) Apply a rotation so that the x axis is collinear with the vector from `holdPoint` to `destinationPoint`. This transformation makes the scaled `pullPoint` move to `destinationPoint`.
5) Apply a translation so that `holdPoint` is back where it was at the beginning of the sequence.

The idea is that we're "stretching" the pull point to the destination, and the object we're stretching is flexible. This means that we expect a shear to happen with this transformation, so we do a rotation + non-uniform scale + rotation to simulate that. I think non-uniform scaling is useful to have so we can do things like stretch spheres into ellipsoids, so this function and `scaleByFactor` should do two different things. The rotations also ensure that the pull and destination points don't have to be collinear, which might be flexibility we want to support.

File: src/geometry/WorkingGeometry.ts (hunk 3 above)
Comment (abhimadan) [reply src/geometry/WorkingGeometry.ts]: I removed it because I misread what was going on here before as using the distance as part of the scaling factor, which can produce incorrect results like zero-scaling along some axes. However, if we want this as a "rigid body" counterpart to the modified scale function, it would make sense to add that back, so I'll do that.

File: tests/geometry/WorkingGeometry.spec.ts (hunk 4 above)
Comment (abhimadan) [reply tests/geometry/WorkingGeometry.spec.ts]: (I'm omitting the z coordinate for clarity)
So this is holding the bottom-left corner of the square and pulling the bottom-right corner to (1,1), while allowing the square to be skewed in the process. However, after translating the bottom-left corner to be the origin, the pull point becomes (-1, 0) and the destination point becomes (0, 1), and the entire function resolves to becoming a -90 degree rotation about (1,0). Does that make sense?

Though this is a good point to bring up, since calling this function `scale` doesn't make much sense. It might be better to rename the pair of scaling functions (based on the previous two threads) as follows:

- `scale` -> `stretchTo`
- `scaleByFactor` -> `rigidStretchTo`

What do you think?

Comment (abhimadan) [reply tests/geometry/WorkingGeometry.spec.ts]: Err, maybe s/`rigidStretchTo`/`rigidPullTo`? That could make it a bit clearer that there's no non-uniform scaling going on in that function.

Comment (tammy) [reply tests/geometry/WorkingGeometry.spec.ts]: Drew this out as well and it made sense! I mixed up the ordering for the `scale` function parameters initially so we're all good 😄 

I also think we should rename the functions. In 2D I think of them akin to stretching a shape in freeform vs. stretching it with constrained proportions. For example I'll think "they're the same operation but one is just a constraint" but with `rigidPullTo` it sounds like a different operation to me. Something like `stretch` and `stretchProportionally` makes more sense in my mind. This is super subjective so thoughts?

Comment (abhimadan) [reply tests/geometry/WorkingGeometry.spec.ts]: I think the distinction between "freeform" vs "proportional" stretching is good! How about `freeformStretchTo` and `proportionalStretchTo`? That way the names are consistently structured, a user can see the similarities and differences between the two functions.

Comment (tammy) [reply tests/geometry/WorkingGeometry.spec.ts]: That sounds good! What do you think about possibly changing the ordering of the words? I'm just thinking that it's easier to search up `stretchFreeformTo` and `stretchProportionalTo` since they start with the same letters. But that's a nit which we can easily change later, and I'm also not a fan of the `stretchPropotionalTo` since I think the meaning is distorted. 🤔

File: src/geometry/WorkingGeometry.ts (hunk 3 above)
Comment (abhimadan) [reply src/geometry/WorkingGeometry.ts]: Ah never mind, I think adding back `pullPoint` still isn't necessary because this is uniform scaling, so it scales equally in all directions — we don't need to specify a specific direction. The only thing that needs to be specified is where the scale originates from, which is `holdPoint`. We might want to add a rotation here like in the other function, but then we would also need to specify a `destinationPoint` to determine the rotation direction. Since the scaling factor is specified directly, that `destinationPoint` might not actually be the real destination point, so if we decide to add a `destinationPoint`, we should remove `factor`.

So basically I think there are two options here:
1) Leave the interface as is. A user can't specify a rotation directly in this function, but it's easy to understand and use.
2) Use the same interface as the other function. This makes it more complex, but then the difference between the two functions is more clear.

I'm personally leaning towards option 1 (and perhaps using a similar interface for the other function as well, but I'm not sure if that would make it easier to use, so I won't do it in this PR regardless). Thoughts?

File: tests/geometry/WorkingGeometry.spec.ts (hunk 4 above)
Comment (abhimadan) [reply tests/geometry/WorkingGeometry.spec.ts]: I think `.*StretchTo` should still be easy to search for, right? If it isn't, we could file an umbrella issue for online docs and add this search grouping as a feature we want.

File: src/geometry/WorkingGeometry.ts [hunk 5]
```diff
@@ -149,45 +154,77 @@ export class WorkingGeometry {
...
      * @param {vec3} holdPoint: point to scale from, default to true origin.
      */
-    public scaleByFactor(factor: number, pullPoint: vec3, holdPoint: vec3 = vec3.create()) {
-        const scalingDirection: vec3 = vec3.sub(vec3.create(), pullPoint, holdPoint);
-        const absScalingDirection: vec3 = vec3.fromValues(
-            Math.abs(scalingDirection[0]),
-            Math.abs(scalingDirection[1]),
-            Math.abs(scalingDirection[2])
-        );
+    public proportionalStretchTo(factor: number, holdPoint: vec3 = vec3.create()) {
```
Comment (pbardea) [inline src/geometry/WorkingGeometry.ts]: So I may have missed part of a conversation, but it looks like this function removed the pull point. The pull point allowed for scaling by a given factor on a given axis. It looks like proportionalStetchTo scales the object in every direction, which has its uses, but I can see the merits of keeping a function which can scale along a given axis by a certain factor.

File: src/geometry/WorkingGeometry.ts [hunk 6]
```diff
@@ -279,5 +261,13 @@ export class WorkingGeometry {
         this.vertices = this.vertices.map((workingVec: vec4) =>
```
Comment (pbardea) [inline src/geometry/WorkingGeometry.ts]: Nit: can we capitalize "transformation matrix" in the docstring above?

File: tests/geometry/WorkingGeometry.spec.ts (hunk 4 above)
Comment (pbardea) [reply tests/geometry/WorkingGeometry.spec.ts]: Personal opinion: if we ever implement some form of auto completion in an editor putting all stretching functions with the same prefix would be useful. Also since our target audience is people who are not familiar with programming environments, I would prefer to not have them to reach for regexs 😄

**Review by pbardea — APPROVED** [review]
LGTM, but I think it would be useful to add back a method which can scale by a factor on a given axis. Thanks for doing this! 🎉

File: tests/geometry/WorkingGeometry.spec.ts (hunk 4 above)
Comment (tammy) [reply tests/geometry/WorkingGeometry.spec.ts]: @abhimadan did you mean that the user will type in `.*StretchTo`? I'm unfamiliar with regexes myself but I would imagine the user to start typing `stretch` and then expect to see the functions they're looking for. As long as they're using an editor that does fuzzy search typeahead suggestions then it shouldn't be a problem :)

File: src/geometry/WorkingGeometry.ts (hunk 5 above)
Comment (tammy) [reply src/geometry/WorkingGeometry.ts]: @abhimadan You mentioned that it "scales equally in all directions", does that mean the `holdPoint` of the object will no longer have the coordinates of `holdPoint` after scaling? It sounds like `holdPoint` becomes the "centre" of scaling whereas the expected behaviour is for it to be the single unmoving point.

Comment (abhimadan) [reply src/geometry/WorkingGeometry.ts]: That's what `freefromStretchTo` is for :)

That function doesn't allow you to specify a scale by an explicit factor, but it does scale along an axis.

Comment (tammy) [reply src/geometry/WorkingGeometry.ts]: I also got confused with a few things in my previous comments. The purpose of `scale` and `scaleByFactor` were to allow the user to do the same things but with different interfaces.
The purpose for each:
* With `scale` it's possible for the user to freeform alter the object however they like. If they want it to look a certain way they can specify a coordinate value and then scale their object. They can also skew their object using this function.
* With `scaleByFactor` it'll allow for them to make accurate changes to their model if they very much require units that are precise. The user should not be able to skew their function since the direction of the scale is always `pullPoint - holdPoint`, but otherwise it's the same as `scale`. At the time of the initial PR we decided to do skewing at a later time.

I think I got confused with the proportional scaling since our test only had that, but the initial purpose was not to make `scaleByFactor` be proportionally constrained. I was actually going to use `scale` inside `scaleByFactor` but because there was too much calculations and then undoing I decided otherwise.

Comment (pbardea) [reply src/geometry/WorkingGeometry.ts]: Right, but I think it would also be useful if the user could specify a stretch along a diagonal axis say, without needing to do the math themselves.

File: src/geometry/WorkingGeometry.ts (hunk 6 above)
Comment (abhimadan) [reply src/geometry/WorkingGeometry.ts]: Do you mean in the error? I don't see a docstring that has it. If that's what you mean, I'm OK with changing that (provided we use the same convention with other error strings).

File: tests/geometry/WorkingGeometry.spec.ts (hunk 4 above)
Comment (abhimadan) [reply tests/geometry/WorkingGeometry.spec.ts]: (So `.` is a wildcard character in most regex engines, and `*` matches the preceeding character zero or more times, so `.*` basically means any string.)

My bad, I was thinking of users searching through an online docs page, not about editor autocompletion.  In the former case, I don't think implicitly doing a `.*[searchtext].*` on every search would be a problem, and I'm not familiar enough with how editor autocomplete plugins work to comment on the latter. I don't want to commit too much to names right now since it should be simple enough to change them later, but I'm not a huge fan of the `stretch.*To` names as alternatives since they seem to distort the meaning of the names a bit.

File: src/geometry/WorkingGeometry.ts (hunk 5 above)
Comment (abhimadan) [reply src/geometry/WorkingGeometry.ts]: Sorry about all the churn on this PR — I didn't follow the PR that brought in these changes, and the original implementation didn't seem correct to me because it could produce uninvertible matrices (which would mess up normal transformations).

So if we want non-uniform scaling here, the implementation will almost be the same as `freeformStretchTo`, except that step 3 (see the earlier comment explaining the implementation) will use the provided factor as the x-scale, and step 4 will rotate back to the same line as `pullPoint`. Is that OK with you @taimengliu and @pbardea?

Comment (abhimadan) [reply src/geometry/WorkingGeometry.ts]: FWIW, I still think that having a uniform scaling option is useful, in case you don't want to shear your object.

Comment (pbardea) [reply src/geometry/WorkingGeometry.ts]: +1 i can see both of them being useful.

File: src/geometry/WorkingGeometry.ts (hunk 6 above)
Comment (pbardea) [reply src/geometry/WorkingGeometry.ts]: I can't comment on that line since it was from an old PR and I just saw it while browsing this one. It's on like 258 (formerly 276) of this file. I'm ok w/ not changing it, and I'll just follow up with it in a different PR.

**Review by tammy — APPROVED** [review]

**Comment by tammy** [comment]
Thanks for talking through everything! We're GTG 🎉 💯 😸

**Comment by abhimadan** [comment]
Thanks for reviewing this everyone! I added a `freeformStretchByFactor` that works as discussed. I'll merge this PR now.
